doc/tutorials/03_json.md

```diff
@@ -0,0 +1,184 @@
+case class Meal(name: String, servings: Int, ingredients: List[String])
+case class Nutrition(name: String, healthy: Boolean, calories: Int)
+```
+
+The first step is to define the functions that will enable the serialization and deserialization of these classes to JSON. This
+can be done by hand, but most of the time, we can rely on derivation: a compile-time process that generates the
+code needed to transform a `String` into a `Meal` (or an error) and to transform a `Nutrition` into a `String.`
+
+This is the task of our chosen JSON library. By adding a `... derives` clause, an instance of a `JsonValueCodec` will
+be generated at compile-time (with compile-time errors if the library can't figure out how to serialize/deserialize some
+component).
+
+We can also test the serialization of an example object. Let's put this in a `json.scala` file:
+
+```scala
+//> using dep com.softwaremill.sttp.tapir::tapir-jsoniter-scala:@VERSION@
+//> using dep com.github.plokhotnyuk.jsoniter-scala::jsoniter-scala-macros:2.29.0
+
+import com.github.plokhotnyuk.jsoniter_scala.core.* // needed for `writeToString`
+import com.github.plokhotnyuk.jsoniter_scala.macros.* // needed for ... derives
+
+case class Meal(name: String, servings: Int, ingredients: List[String]) 
+  derives ConfiguredJsonValueCodec
+case class Nutrition(name: String, healthy: Boolean, calories: Int) 
+  derives ConfiguredJsonValueCodec
+
+@main def tapirJson(): Unit =
+  println(writeToString(Meal("salad", 1, List("lettuce", "tomato", "cucumber"))))
+```
+
+```{note}
+Even though we request derivation of a `ConfiguredJsonValueCodec`, we obtain a `JsonValueCodec` instance. This is due to
+the way jsoniter-scala works; the "configured" variant accepts an implicit `CodecMakerConfig`, which can be used to 
+configure the generation process. 
+```
+
+This should output the following:
+
+```bash
+% scala-cli json.scala
+{"name":"salad","servings":1,"ingredients":["lettuce","tomato","cucumber"]}
+```
+
+## Deriving schema
+
+With the functions translating between JSON strings and our high-level types ready, we can take care of the second
+component: schemas. As mentioned in the beginning, schemas are needed to generate accurate OpenAPI documentation and
+validation. They can be defined by hand, but most of the time, you can use compile-time derivation - just as with
+JSON codecs.
+
+In our case, deriving the schemas will amount to adding a `... derives Schema` clause. Let's run a quick test:
+
+{emphasize-lines="7, 10, 12, 16"}
+```scala
+//> using dep com.softwaremill.sttp.tapir::tapir-jsoniter-scala:@VERSION@
+//> using dep com.github.plokhotnyuk.jsoniter-scala::jsoniter-scala-macros:2.29.0
+
+import com.github.plokhotnyuk.jsoniter_scala.core.*   // needed for `writeToString`
+import com.github.plokhotnyuk.jsoniter_scala.macros.* // needed for ... derives
+
+import sttp.tapir.*
+
+case class Meal(name: String, servings: Int, ingredients: List[String]) 
+  derives ConfiguredJsonValueCodec, Schema
```

Comment:
**question:** Is this the recommended default approach? (Using 'derives'). In adopt-tapir we use `given`s for JSON, and automatic schema derivation. Personally I'm more used to keeping the model classes clean of dependencies and I'd use givens with endpoint definitions, but I suppose you chose `derives` for simplicity?
Also - what do you think about splitting this tutorial and tackling schemas in a separate one, to make it easier to grok? This one could use `import sttp.tapir.generic.auto.*`, and the second one could extend it to introduce the concept of schemas and ways of deriving it.

Reply:
Using `derives` is most simplest. Using `given` is more advanced approach. Then when API evolves through versions the next step is to have 2+ sets of data structures: 1st is for the internal data model and others for API versions that matches JSON structure as much as possible. Transforming libraries like [chimney](https://github.com/scalalandio/chimney) and [ducktape](https://github.com/arainko/ducktape) help to reduce boilerplate when evolving the internal data model and new versions of API.

Reply:
@kciesielski I think the least magical and simples approach is to use `... derives`. There's no special imports, and I think it's easier to understand where these instances come from (since they are explicitly requested). If anything, maybe we should change the defaults in adopt-tapir. As @plokhotnyuk writes, as your project matures, you'll want to split the derivation/data definitions, but that's far beyond a tutorial :)

Reply:
I'm going to follow [The Grand Unified Theory of Documentation](https://docs.divio.com/documentation-system/) for writing of docs for jsoniter-scala :) and it seems that this approach is already picked by [tupson](https://sake92.github.io/tupson/).

Reply:
@plokhotnyuk Yes that's also one of my goals, to start separating "reference" documentation from "tutorials". I didn't plan "how-to" guides initially, but after some comments on the tutorials this might be a good idea to include them as well.